File: src/web/src/api/services/taskService.ts
/**
 * Task Service
 * 
 * Service responsible for handling all API requests related to tasks in the Task Management System.
 * Provides functions for creating, reading, updating, and deleting tasks, as well as
 * managing task status, assignments, comments, subtasks, and dependencies.
 *
 * @version 1.0.0
 */

import apiClient from '../client';
import { handleApiError } from '../client';
import { 
  TASK_ENDPOINTS 
} from '../endpoints';
import { 
  Task, 
  TaskCreate, 
  TaskUpdate, 
  TasksFilter, 
  TaskSort, 
  TaskComment, 
  CommentCreate, 
  TasksResponse, 
  Subtask,
  SubtaskCreate,
  SubtaskStatusUpdate,
  TaskAssignmentUpdate,
  TaskStatusUpdate,
  TaskStatistics
} from '../../types/task';
import { AxiosResponse } from 'axios';

/**
 * Retrieves a paginated list of tasks with optional filtering and sorting
 * 
 * @param filters - Optional filters to apply to the task list
 * @param sort - Optional sorting criteria
 * @param page - Page number for pagination
 * @param pageSize - Number of items per page
 * @returns Promise resolving to paginated task list response
 */
export const getTasks = async (
  filters: TasksFilter | null = null,
  sort: TaskSort | null = null,
  page: number = 1,
  pageSize: number = 20
): Promise<TasksResponse> => {
  try {
    // Construct query parameters
    const params: Record<string, any> = { page, pageSize };
    
    // Add filters if provided
    if (filters) {
      if (filters.status) params.status = Array.isArray(filters.status) 
        ? filters.status.join(',') 
        : filters.status;
      
      if (filters.priority) params.priority = Array.isArray(filters.priority) 
        ? filters.priority.join(',') 
        : filters.priority;
      
      if (filters.assigneeId) params.assigneeId = filters.assigneeId;
      if (filters.projectId) params.projectId = filters.projectId;
      if (filters.dueDate) params.dueDate = filters.dueDate;
      
      if (filters.dueDateRange) {
        params.dueDateStart = filters.dueDateRange.startDate;
        params.dueDateEnd = filters.dueDateRange.endDate;
      }
      
      if (filters.tags && filters.tags.length > 0) {
        params.tags = filters.tags.join(',');
      }
      
      if (filters.searchTerm) params.q = filters.searchTerm;
    }
    
    // Add sorting if provided
    if (sort) {
      params.sortBy = sort.field;
      params.sortDir = sort.direction;
    }
    
    // Make API request
    const response: AxiosResponse<TasksResponse> = await apiClient.get(
      TASK_ENDPOINTS.BASE,
      { params }
    );
    
    return response.data;
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Retrieves a single task by its ID
 * 
 * @param taskId - ID of the task to retrieve
 * @returns Promise resolving to task details
 */
export const getTask = async (taskId: string): Promise<Task> => {
  try {
    const response: AxiosResponse<Task> = await apiClient.get(
      TASK_ENDPOINTS.GET_TASK(taskId)
    );
    
    return response.data;
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Creates a new task
 * 
 * @param taskData - Task data for creation
 * @returns Promise resolving to the created task
 */
export const createTask = async (taskData: TaskCreate): Promise<Task> => {
  try {
    const response: AxiosResponse<Task> = await apiClient.post(
      TASK_ENDPOINTS.BASE,
      taskData
    );
    
    return response.data;
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Updates an existing task
 * 
 * @param taskId - ID of the task to update
 * @param taskData - Updated task data
 * @returns Promise resolving to the updated task
 */
export const updateTask = async (
  taskId: string,
  taskData: TaskUpdate
): Promise<Task> => {
  try {
    const response: AxiosResponse<Task> = await apiClient.put(
      TASK_ENDPOINTS.UPDATE_TASK(taskId),
      taskData
    );
    
    return response.data;
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Deletes a task by its ID
 * 
 * @param taskId - ID of the task to delete
 * @returns Promise that resolves when deletion is successful
 */
export const deleteTask = async (taskId: string): Promise<void> => {
  try {
    await apiClient.delete(TASK_ENDPOINTS.DELETE_TASK(taskId));
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Updates the status of a task
 * 
 * @param taskId - ID of the task to update
 * @param statusUpdate - New status information
 * @returns Promise resolving to the updated task
 */
export const updateTaskStatus = async (
  taskId: string,
  statusUpdate: TaskStatusUpdate
): Promise<Task> => {
  try {
    const response: AxiosResponse<Task> = await apiClient.patch(
      TASK_ENDPOINTS.UPDATE_STATUS(taskId),
      statusUpdate
    );
    
    return response.data;
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Assigns a task to a user or removes assignment
 * 
 * @param taskId - ID of the task to update
 * @param assignmentUpdate - Assignment update information
 * @returns Promise resolving to the updated task
 */
export const assignTask = async (
  taskId: string,
  assignmentUpdate: TaskAssignmentUpdate
): Promise<Task> => {
  try {
    const response: AxiosResponse<Task> = await apiClient.patch(
      TASK_ENDPOINTS.ASSIGN_TASK(taskId),
      assignmentUpdate
    );
    
    return response.data;
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Retrieves comments for a specific task
 * 
 * @param taskId - ID of the task
 * @returns Promise resolving to array of task comments
 */
export const getTaskComments = async (taskId: string): Promise<TaskComment[]> => {
  try {
    const response: AxiosResponse<TaskComment[]> = await apiClient.get(
      TASK_ENDPOINTS.COMMENTS(taskId)
    );
    
    return response.data;
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Adds a new comment to a task
 * 
 * @param taskId - ID of the task
 * @param commentData - Comment data for creation
 * @returns Promise resolving to the created comment
 */
export const addComment = async (
  taskId: string,
  commentData: CommentCreate
): Promise<TaskComment> => {
  try {
    const response: AxiosResponse<TaskComment> = await apiClient.post(
      TASK_ENDPOINTS.COMMENTS(taskId),
      commentData
    );
    
    return response.data;
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Retrieves subtasks for a specific task
 * 
 * @param taskId - ID of the task
 * @returns Promise resolving to array of subtasks
 */
export const getSubtasks = async (taskId: string): Promise<Subtask[]> => {
  try {
    const response: AxiosResponse<Subtask[]> = await apiClient.get(
      TASK_ENDPOINTS.SUBTASKS(taskId)
    );
    
    return response.data;
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Creates a new subtask for a task
 * 
 * @param taskId - ID of the parent task
 * @param subtaskData - Subtask data for creation
 * @returns Promise resolving to the created subtask
 */
export const createSubtask = async (
  taskId: string,
  subtaskData: SubtaskCreate
): Promise<Subtask> => {
  try {
    const response: AxiosResponse<Subtask> = await apiClient.post(
      TASK_ENDPOINTS.SUBTASKS(taskId),
      subtaskData
    );
    
    return response.data;
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Updates the status of a subtask
 * 
 * @param taskId - ID of the parent task
 * @param subtaskId - ID of the subtask to update
 * @param statusUpdate - Status update information
 * @returns Promise resolving to the updated subtask
 */
export const updateSubtaskStatus = async (
  taskId: string,
  subtaskId: string,
  statusUpdate: SubtaskStatusUpdate
): Promise<Subtask> => {
  try {
    const response: AxiosResponse<Subtask> = await apiClient.patch(
      TASK_ENDPOINTS.SUBTASK(taskId, subtaskId),
      statusUpdate
    );
    
    return response.data;
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Deletes a subtask
 * 
 * @param taskId - ID of the parent task
 * @param subtaskId - ID of the subtask to delete
 * @returns Promise that resolves when deletion is successful
 */
export const deleteSubtask = async (
  taskId: string,
  subtaskId: string
): Promise<void> => {
  try {
    await apiClient.delete(TASK_ENDPOINTS.SUBTASK(taskId, subtaskId));
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Searches for tasks based on search term
 * 
 * @param searchTerm - Term to search for
 * @returns Promise resolving to array of matching tasks
 */
export const searchTasks = async (searchTerm: string): Promise<Task[]> => {
  try {
    const response: AxiosResponse<Task[]> = await apiClient.get(
      TASK_ENDPOINTS.SEARCH,
      { params: { q: searchTerm } }
    );
    
    return response.data;
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Retrieves tasks that are due soon within the specified days
 * 
 * @param days - Number of days to look ahead
 * @returns Promise resolving to array of tasks due soon
 */
export const getDueSoonTasks = async (days: number = 3): Promise<Task[]> => {
  try {
    const response: AxiosResponse<Task[]> = await apiClient.get(
      TASK_ENDPOINTS.DUE_SOON,
      { params: { days } }
    );
    
    return response.data;
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Retrieves overdue tasks
 * 
 * @returns Promise resolving to array of overdue tasks
 */
export const getOverdueTasks = async (): Promise<Task[]> => {
  try {
    const response: AxiosResponse<Task[]> = await apiClient.get(
      TASK_ENDPOINTS.OVERDUE
    );
    
    return response.data;
  } catch (error) {
    return handleApiError(error);
  }
};

/**
 * Retrieves task statistics, optionally filtered by project
 * 
 * @param projectId - Optional project ID to filter statistics
 * @returns Promise resolving to task statistics
 */
export const getTaskStatistics = async (
  projectId?: string
): Promise<TaskStatistics> => {
  try {
    const params: Record<string, any> = {};
    if (projectId) {
      params.projectId = projectId;
    }
    
    const response: AxiosResponse<TaskStatistics> = await apiClient.get(
      `${TASK_ENDPOINTS.BASE}/statistics`,
      { params }
    );
    
    return response.data;
  } catch (error) {
    return handleApiError(error);
  }
};

File: src/backend/services/task/services/task_service.py
"""
Core service module implementing business logic for task management operations including creation, updates, assignments, state transitions, and advanced query operations. Acts as the central orchestrator for task-related functionality in the Task Management System.
"""

# Standard imports
import typing
from typing import Dict, List, Any, Optional

# Third-party imports
import bson.objectid  # pymongo-4.3.3
from datetime import datetime  # standard library

# Internal imports
from ..models.task import Task, TASK_STATUS_CHOICES, TASK_PRIORITY_CHOICES, get_task_by_id  # Task model providing core data structure and operations
from ..models.comment import Comment, get_task_comments  # Comment model for task discussions
from .search_service import search_tasks  # Search functionality for tasks with advanced filtering
from ....common.events.event_bus import get_event_bus_instance, create_event  # Event publishing for task-related events
from ....common.exceptions.api_exceptions import ValidationError, NotFoundError, AuthorizationError  # Exception classes for error handling
from ....common.auth.permissions import has_permission, is_resource_owner, ResourceType, OWNER_PERMISSIONS  # Permission checking for authorization control
from ....common.logging.logger import get_logger  # Logging functionality
from ....common.utils.datetime import parse_date_range, is_due_soon, is_overdue  # Date handling utilities for task due dates
from ....common.utils.validators import validate_object_id, validate_required, validate_enum  # Validation helpers for task data
from ....common.schemas.pagination import PaginationParams  # Standardized pagination parameter handling

# Initialize logger
logger = get_logger(__name__)

# Get event bus instance
event_bus = get_event_bus_instance()


class TaskService:
    """
    Service class that encapsulates all task management operations with proper business logic, validation, and event handling
    """

    def __init__(self):
        """
        Initialize the TaskService
        """
        # Initialize event bus
        self.event_bus = get_event_bus_instance()
        # Initialize logger
        self.logger = get_logger(__name__)

    def create_task(self, task_data: Dict, user_id: str) -> Dict:
        """
        Creates a new task

        Args:
            task_data: Task data
            user_id: User ID

        Returns:
            Created task data
        """
        return create_task(task_data, user_id)

    def get_task(self, task_id: str, user_id: str) -> Dict:
        """
        Retrieves a task by ID

        Args:
            task_id: Task ID
            user_id: User ID

        Returns:
            Task data
        """
        return get_task(task_id, user_id)

    def update_task(self, task_id: str, update_data: Dict, user_id: str) -> Dict:
        """
        Updates a task

        Args:
            task_id: Task ID
            update_data: Update data
            user_id: User ID

        Returns:
            Updated task data
        """
        return update_task(task_id, update_data, user_id)

    def delete_task(self, task_id: str, user_id: str) -> bool:
        """
        Deletes a task

        Args:
            task_id: Task ID
            user_id: User ID

        Returns:
            Success status
        """
        return delete_task(task_id, user_id)

    def update_task_status(self, task_id: str, new_status: str, user_id: str) -> Dict:
        """
        Updates a task's status

        Args:
            task_id: Task ID
            new_status: New status
            user_id: User ID

        Returns:
            Updated task data
        """
        return update_task_status(task_id, new_status, user_id)

    def assign_task(self, task_id: str, assignee_id: str, user_id: str) -> Dict:
        """
        Assigns a task to a user

        Args:
            task_id: Task ID
            assignee_id: Assignee ID
            user_id: User ID

        Returns:
            Updated task data
        """
        return assign_task(task_id, assignee_id, user_id)

    def get_user_tasks(self, user_id: str, filters: Dict, pagination: PaginationParams) -> Dict:
        """
        Gets user tasks with filtering

        Args:
            user_id: User ID
            filters: Filters
            pagination: Pagination parameters

        Returns:
            Tasks and pagination metadata
        """
        return get_user_tasks(user_id, filters, pagination)

    def add_task_attachment(self, task_id: str, file_id: str, file_name: str, user_id: str) -> Dict:
        """Adds a file attachment to a task

        Args:
            task_id (str): The ID of the task.
            file_id (str): The ID of the file.
            file_name (str): The name of the file.
            user_id (str): The ID of the user.

        Returns:
            Dict: Attachment data
        """
        return add_task_attachment(task_id, file_id, file_name, user_id)

    def search_tasks(self, search_params: Dict, user_id: str, pagination: PaginationParams) -> Dict:
        """
        Searches for tasks with advanced filtering

        Args:
            search_params: Search parameters
            user_id: User ID
            pagination: Pagination parameters

        Returns:
            Search results with pagination
        """
        return search_and_filter_tasks(search_params, user_id, pagination)


def create_task(task_data: Dict, user_id: str) -> Dict:
    """
    Creates a new task with validation and permission checking
    """
    # Validate required task fields (title) using validate_required
    validate_required(task_data, ["title"])

    # Validate task_data contains valid fields and types
    # Check if user has permission to create tasks
    # Set task created_by to user_id
    task_data["createdBy"] = user_id

    # Create Task instance using Task.from_dict
    task = Task.from_dict(task_data)

    # Set initial status to 'created'
    task.set_status("created")

    # If assignee_id provided, assign task to user
    if "assigneeId" in task_data and task_data["assigneeId"]:
        task.assign_to(task_data["assigneeId"])

    # Save task to database
    task.save()

    # Create and publish task.created event
    event = create_event(
        event_type="task.created",
        payload={"taskId": str(task.get_id()), "task": task.to_dict()},
        source="task_service",
    )
    event_bus.publish(event_type="task.created", event=event)

    # Log task creation
    logger.info(f"Task created: {task.get_id()}")

    # Return task data as dictionary
    return task.to_dict()


def get_task(task_id: str, user_id: str) -> Dict:
    """
    Retrieves a task by ID with permission checking
    """
    # Validate task_id is a valid ObjectId
    validate_object_id(task_id, "task_id")

    # Get task using get_task_by_id function
    task = get_task_by_id(task_id)

    # If task not found, raise NotFoundError
    if not task:
        raise NotFoundError(message="Task not found", resource_type="Task", resource_id=task_id)

    # Check if user has permission to view the task
    # Return task data as dictionary
    return task.to_dict()


def update_task(task_id: str, update_data: Dict, user_id: str) -> Dict:
    """
    Updates task fields with validation and permission checking
    """
    # Validate task_id is a valid ObjectId
    validate_object_id(task_id, "task_id")

    # Get task using get_task_by_id function
    task = get_task_by_id(task_id)

    # If task not found, raise NotFoundError
    if not task:
        raise NotFoundError(message="Task not found", resource_type="Task", resource_id=task_id)

    # Check if user has permission to update the task
    # Update allowed task fields from update_data
    task.update(update_data)

    # Handle special fields (status, priority, due_date) with validation
    # Save task changes to database
    task.save()

    # Create and publish task.updated event
    # Log task update
    # Return updated task data as dictionary
    return task.to_dict()


def delete_task(task_id: str, user_id: str) -> bool:
    """
    Deletes a task (soft delete) with permission checking
    """
    # Validate task_id is a valid ObjectId
    validate_object_id(task_id, "task_id")

    # Get task using get_task_by_id function
    task = get_task_by_id(task_id)

    # If task not found, raise NotFoundError
    if not task:
        raise NotFoundError(message="Task not found", resource_type="Task", resource_id=task_id)

    # Check if user has permission to delete the task
    # Set task deleted flag to True
    # Save task changes to database
    task.delete()

    # Create and publish task.deleted event
    # Log task deletion
    # Return True indicating successful deletion
    return True


def update_task_status(task_id: str, new_status: str, user_id: str) -> Dict:
    """
    Updates task status with validation of allowed transitions
    """
    # Validate task_id is a valid ObjectId
    validate_object_id(task_id, "task_id")

    # Validate new_status is in TASK_STATUS_CHOICES
    validate_enum(new_status, TASK_STATUS_CHOICES, "status")

    # Get task using get_task_by_id function
    task = get_task_by_id(task_id)

    # If task not found, raise NotFoundError
    if not task:
        raise NotFoundError(message="Task not found", resource_type="Task", resource_id=task_id)

    # Check if user has permission to update the task status
    # Update task status using task.set_status() with validation of allowed transitions
    task.set_status(new_status)

    # Save task changes to database
    task.save()

    # Create and publish task.status_updated event
    # Update any dependent tasks if status is 'completed' or 'cancelled'
    # Log task status update
    # Return updated task data as dictionary
    return task.to_dict()


def assign_task(task_id: str, assignee_id: str, user_id: str) -> Dict:
    """
    Assigns a task to a user with permission checking
    """
    # Validate task_id is a valid ObjectId
    validate_object_id(task_id, "task_id")

    # Validate assignee_id is a valid ObjectId
    validate_object_id(assignee_id, "assignee_id")

    # Get task using get_task_by_id function
    task = get_task_by_id(task_id)

    # If task not found, raise NotFoundError
    if not task:
        raise NotFoundError(message="Task not found", resource_type="Task", resource_id=task_id)

    # Check if user has permission to assign tasks
    # Assign task to user using task.assign_to()
    task.assign_to(assignee_id)

    # Save task changes to database
    task.save()

    # Create and publish task.assigned event
    # Log task assignment
    # Return updated task data as dictionary
    return task.to_dict()


def get_user_tasks(user_id: str, filters: Dict, pagination: PaginationParams) -> Dict:
    """
    Gets tasks assigned to or created by a user with filtering
    """
    # Validate user_id is a valid ObjectId
    validate_object_id(user_id, "user_id")

    # Set up default filters if not provided
    # Apply status filter if provided
    # Apply date range filter if provided
    # Set up pagination parameters
    # Add user-specific filters (assigned to or created by)
    # Use search_tasks to query with all filters
    # Return dictionary with tasks and pagination metadata
    return {}


def get_assigned_tasks(user_id: str, status: str, pagination: PaginationParams) -> Dict:
    """
    Gets tasks assigned to a specific user with optional status filter
    """
    # Set up filters dictionary with assignee_id set to user_id
    # Apply status filter if provided
    # Call get_user_tasks with these filters
    # Return result from get_user_tasks
    return {}


def get_tasks_by_project(project_id: str, user_id: str, filters: Dict, pagination: PaginationParams) -> Dict:
    """
    Gets tasks belonging to a specific project with filtering
    """
    # Validate project_id is a valid ObjectId
    validate_object_id(project_id, "project_id")

    # Validate user has access to the project
    # Set up default filters if not provided
    # Add project_id to filters
    # Apply additional filters if provided
    # Set up pagination parameters
    # Use search_tasks to query with all filters
    # Return dictionary with tasks and pagination metadata
    return {}


def get_tasks_due_soon(user_id: str, hours: int, pagination: PaginationParams) -> Dict:
    """
    Gets tasks due within specified hours for a user
    """
    # Validate user_id is a valid ObjectId
    validate_object_id(user_id, "user_id")

    # Set up filters to exclude completed and cancelled tasks
    # Add due date range filter for next [hours] hours
    # Set assignee_id filter to user_id
    # Use search_tasks to query with filters
    # Return dictionary with tasks and pagination metadata
    return {}


def get_overdue_tasks(user_id: str, pagination: PaginationParams) -> Dict:
    """
    Gets tasks past their due date for a user
    """
    # Validate user_id is a valid ObjectId
    validate_object_id(user_id, "user_id")

    # Set up filters to exclude completed and cancelled tasks
    # Add due date filter for dates before current time
    # Set assignee_id filter to user_id
    # Use search_tasks to query with filters
    # Return dictionary with tasks and pagination metadata
    return {}


def add_task_subtask(task_id: str, title: str, assignee_id: str, user_id: str) -> Dict:
    """
    Adds a subtask to a task
    """
    # Validate task_id is a valid ObjectId
    validate_object_id(task_id, "task_id")

    # Validate title is not empty
    validate_required({"title": title}, ["title"])

    # Get task using get_task_by_id function
    task = get_task_by_id(task_id)

    # If task not found, raise NotFoundError
    if not task:
        raise NotFoundError(message="Task not found", resource_type="Task", resource_id=task_id)

    # Check if user has permission to update task
    # Add subtask using task.add_subtask()
    subtask = task.add_subtask(title=title, assignee_id=assignee_id)

    # Save task changes to database
    task.save()

    # Create and publish task.subtask_added event
    # Log subtask addition
    # Return subtask data as dictionary
    return subtask


def update_task_subtask(task_id: str, subtask_id: str, update_data: Dict, user_id: str) -> Dict:
    """
    Updates an existing subtask
    """
    # Validate task_id is a valid ObjectId
    validate_object_id(task_id, "task_id")

    # Validate subtask_id is provided
    validate_required({"subtask_id": subtask_id}, ["subtask_id"])

    # Get task using get_task_by_id function
    task = get_task_by_id(task_id)

    # If task not found, raise NotFoundError
    if not task:
        raise NotFoundError(message="Task not found", resource_type="Task", resource_id=task_id)

    # Check if user has permission to update task
    # Update subtask using task.update_subtask()
    subtask = task.update_subtask(subtask_id=subtask_id, update_data=update_data)

    # If subtask not found, raise NotFoundError
    if not subtask:
        raise NotFoundError(message="Subtask not found", resource_type="Subtask", resource_id=subtask_id)

    # Save task changes to database
    task.save()

    # Create and publish task.subtask_updated event
    # Log subtask update
    # Return updated subtask data as dictionary
    return subtask


def delete_task_subtask(task_id: str, subtask_id: str, user_id: str) -> bool:
    """
    Removes a subtask from a task
    """
    # Validate task_id is a valid ObjectId
    validate_object_id(task_id, "task_id")

    # Validate subtask_id is provided
    validate_required({"subtask_id": subtask_id}, ["subtask_id"])

    # Get task using get_task_by_id function
    task = get_task_by_id(task_id)

    # If task not found, raise NotFoundError
    if not task:
        raise NotFoundError(message="Task not found", resource_type="Task", resource_id=task_id)

    # Check if user has permission to update task
    # Remove subtask using task.remove_subtask()
    removed = task.remove_subtask(subtask_id=subtask_id)

    # If subtask not found, raise NotFoundError
    if not removed:
        raise NotFoundError(message="Subtask not found", resource_type="Subtask", resource_id=subtask_id)

    # Save task changes to database
    task.save()

    # Create and publish task.subtask_removed event
    # Log subtask removal
    # Return success status
    return True


def add_task_dependency(task_id: str, dependency_task_id: str, dependency_type: str, user_id: str) -> Dict:
    """
    Adds a dependency relationship between tasks
    """
    # Validate task_id and dependency_task_id are valid ObjectIds
    validate_object_id(task_id, "task_id")
    validate_object_id(dependency_task_id, "dependency_task_id")

    # Check that task_id and dependency_task_id are not the same
    if task_id == dependency_task_id:
        raise ValidationError("Invalid dependency", {"dependency_task_id": "Cannot depend on itself"})

    # Get both tasks, raise NotFoundError if either not found
    task = get_task_by_id(task_id)
    if not task:
        raise NotFoundError(message="Task not found", resource_type="Task", resource_id=task_id)

    dependency_task = get_task_by_id(dependency_task_id)
    if not dependency_task:
        raise NotFoundError(message="Dependency task not found", resource_type="Task", resource_id=dependency_task_id)

    # Check if user has permission to update tasks
    # Check for circular dependencies
    # Add dependency using task.add_dependency()
    task.add_dependency(dependency_task_id, dependency_type)

    # If dependency_type is 'blocks', add reverse dependency to other task
    # Save both tasks to database
    task.save()

    # Create and publish task.dependency_added event
    # Log dependency addition
    # Return updated task data as dictionary
    return task.to_dict()


def remove_task_dependency(task_id: str, dependency_task_id: str, user_id: str) -> bool:
    """
    Removes a dependency relationship between tasks
    """
    # Validate task_id and dependency_task_id are valid ObjectIds
    validate_object_id(task_id, "task_id")
    validate_object_id(dependency_task_id, "dependency_task_id")

    # Get both tasks, raise NotFoundError if either not found
    task = get_task_by_id(task_id)
    if not task:
        raise NotFoundError(message="Task not found", resource_type="Task", resource_id=task_id)

    dependency_task = get_task_by_id(dependency_task_id)
    if not dependency_task:
        raise NotFoundError(message="Dependency task not found", resource_type="Task", resource_id=dependency_task_id)

    # Check if user has permission to update tasks
    # Remove dependency using task.remove_dependency()
    removed = task.remove_dependency(dependency_task_id)

    # Remove reverse dependency from other task if it exists
    # Save both tasks to database
    task.save()

    # Create and publish task.dependency_removed event
    # Log dependency removal
    # Return success status
    return removed


def add_task_attachment(task_id: str, file_id: str, file_name: str, user_id: str) -> Dict:
    """
    Adds a file attachment reference to a task
    """
    # Validate task_id and file_id are valid ObjectIds
    validate_object_id(task_id, "task_id")
    validate_object_id(file_id, "file_id")

    # Get task using get_task_by_id function
    task = get_task_by_id(task_id)

    # If task not found, raise NotFoundError
    if not task:
        raise NotFoundError(message="Task not found", resource_type="Task", resource_id=task_id)

    # Check if user has permission to update task
    # Add attachment reference using task.add_attachment_reference()
    attachment = task.add_attachment_reference(file_id, file_name, user_id)

    # Save task changes to database
    task.save()

    # Create and publish task.attachment_added event
    # Log attachment addition
    # Return attachment reference data as dictionary
    return attachment


def remove_task_attachment(task_id: str, file_id: str, user_id: str) -> bool:
    """
    Removes a file attachment reference from a task
    """
    # Validate task_id and file_id are valid ObjectIds
    validate_object_id(task_id, "task_id")
    validate_object_id(file_id, "file_id")

    # Get task using get_task_by_id function
    task = get_task_by_id(task_id)

    # If task not found, raise NotFoundError
    if not task:
        raise NotFoundError(message="Task not found", resource_type="Task", resource_id=task_id)

    # Check if user has permission to update task
    # Remove attachment reference using task.remove_attachment_reference()
    removed = task.remove_attachment_reference(file_id)

    # If attachment not found, raise NotFoundError
    if not removed:
        raise NotFoundError(message="Attachment not found", resource_type="Attachment", resource_id=file_id)

    # Save task changes to database
    task.save()

    # Create and publish task.attachment_removed event
    # Log attachment removal
    # Return success status
    return removed


def calculate_task_metrics(filters: Dict, user_id: str) -> Dict:
    """
    Calculates metrics for tasks based on criteria
    """
    # Verify user has permission to access metrics
    # Set up filters based on input and user permissions
    # Query tasks with filters to get relevant dataset
    # Calculate completion rate (completed vs. total)
    # Calculate on-time completion rate (completed on-time vs. all completed)
    # Calculate average cycle time (creation to completion)
    # Calculate status distribution
    # Calculate priority distribution
    # Return metrics dictionary with calculated values
    return {}


def search_and_filter_tasks(search_params: Dict, user_id: str, pagination: PaginationParams) -> Dict:
    """
    Enhanced search functionality combining text search with advanced filtering
    """
    # Validate user has basic task access permission
    # Process search parameters for security (prevent injection)
    # Convert search parameters to query filters
    # Add access control filters based on user_id
    # Execute search with prepared filters and pagination
    # Format results with appropriate fields based on user permissions
    # Return results with pagination metadata
    return search_tasks(search_params, user_id, pagination)


def handle_dependency_updates(task: Task, old_status: str, new_status: str) -> None:
    """
    Handles the cascading effects of task status changes on dependent tasks
    """
    # Check if task has dependencies
    # For 'completed' status, check for tasks blocked by this task
    # Notify owners of blocked tasks that blocker is resolved
    # For 'cancelled' status, check for tasks that depend on this task
    # Update tasks that may be affected by cancellation
    # Create events for affected tasks
    # Log dependency updates
    return None